### PR TITLE
HIVE-27948: Wrong results when using materialized views with non-deterministic/dynamic functions

### PR DESCRIPTION
**Pull request by kasakrisz**
### What changes were proposed in this pull request?
1. Collect available rewrite algorithms when validating materialized view definition query plan.
If none of the algorithms selected do not register the MV for automatic rewrites.
2. Print warning message to console when enabling MV for rewrite but definition query has unsupported parts for algebraic rewrite algorithm.
3. If MV definition has non-deterministic function call none of the rewrite algorithms are selected.

### Why are the changes needed?
Enabling MVs having non-deterministic functions in automatic query rewrites leads to data correctness issues. See jira for details: https://issues.apache.org/jira/browse/HIVE-27948

### Does this PR introduce _any_ user-facing change?
Yes, query results and explain plan may change.

### Is the change a dependency upgrade?
No.

### How was this patch tested?
```
mvn test -Dtest.output.overwrite -DskipSparkTests -Dtest=TestMiniLlapLocalCliDriver -Dqfile=materialized_view_no_cbo_rewrite.q,materialized_view_rewrite_by_text_5.q,materialized_view_create_rewrite_6.q,materialized_view_rewrite_3.q,materialized_view_rewrite_by_text_10.q -pl itests/qtest -Pitests
```

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/view/materialized/alter/rewrite/AlterMaterializedViewRewriteOperation.java
Comment (zabetak) [inline ql/src/java/org/apache/hadoop/hive/ql/ddl/view/materialized/alter/rewrite/AlterMaterializedViewRewriteOperation.java:74]: There seems to be some inconsistency between CREATE and ALTER VIEW statements. Here we throw an exception when there are no rewrite algorithms but in `SemanticAnalyzer#validateCreateView` we simply log an error to the console.

File: ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java
Comment (zabetak) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java:14672]: Why don't we check the supported algorithms here?

File: ql/src/test/org/apache/hadoop/hive/ql/metadata/TestMaterializedViewsCache.java
Comment (zabetak) [inline ql/src/test/org/apache/hadoop/hive/ql/metadata/TestMaterializedViewsCache.java:160]: Should we use RewriteAlgorithm.ALL instead?

File: ql/src/test/results/clientpositive/llap/materialized_view_no_cbo_rewrite.q.out
Comment (zabetak) [inline ql/src/test/results/clientpositive/llap/materialized_view_no_cbo_rewrite.q.out:39]: It seems that with this patch we also enable more rewritting capabilities that we didn't support before. Maybe worth mentioning in the JIRA ticket as well.

File: ql/src/test/queries/clientpositive/materialized_view_rewrite_by_text_10.q
Comment (zabetak) [inline ql/src/test/queries/clientpositive/materialized_view_rewrite_by_text_10.q:7]: Shouldn't the CREATE fail completely? I mentioned this in the code as well but I don't understand why CREATE and ALTER behave differently.

File: ql/src/test/queries/clientpositive/materialized_view_rewrite_by_text_11.q
Comment (zabetak) [inline ql/src/test/queries/clientpositive/materialized_view_rewrite_by_text_11.q:13]: This appears as a new file although in fact it is a negative test that was moved. Did you try to use `git mv` for performing the change so that it appears better in git history/diff.

File: ql/src/test/queries/clientpositive/materialized_view_rewrite_by_text_10.q
Comment (zabetak) [reply ql/src/test/queries/clientpositive/materialized_view_rewrite_by_text_10.q:7]: Actually thinking a bit more on this, it may be fine to allow the creation of the materialized view as long as it does not create issues with query rewriting and (incremental) refresh.

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/HiveRelOptMaterializationValidator.java
Comment (kasakrisz) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/HiveRelOptMaterializationValidator.java:326]: There are two types of failures:
1. disable both rewrite algs.
2. allow text base only
If type 1) is found we can stop traversing the tree.
If type 2) is found we record it if this is the first one from this type but we have to continue traversing because there might be a type 1).

Actually we print only one failure and if there are more than one from type 2) it doesn't matter which one is printed. I kept the logic avoid too many golden file changes.

File: ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java
Comment (kasakrisz) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java]: Renamed and disable rewrite if no rewrite alg is available.

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/HiveRelOptMaterializationValidator.java
Comment (kasakrisz) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/HiveRelOptMaterializationValidator.java]: done

File: ql/src/java/org/apache/hadoop/hive/ql/metadata/AutomaticRewritingValidationResult.java
Comment (kasakrisz) [reply ql/src/java/org/apache/hadoop/hive/ql/metadata/AutomaticRewritingValidationResult.java]: Removed javadoc and rename the class to MaterializationValidationResult

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/view/materialized/alter/rewrite/AlterMaterializedViewRewriteOperation.java
Comment (kasakrisz) [reply ql/src/java/org/apache/hadoop/hive/ql/ddl/view/materialized/alter/rewrite/AlterMaterializedViewRewriteOperation.java:74]: Changed that create mv not just log the error but also disables rewrite for MV.

File: ql/src/test/org/apache/hadoop/hive/ql/metadata/TestMaterializedViewsCache.java
Comment (kasakrisz) [reply ql/src/test/org/apache/hadoop/hive/ql/metadata/TestMaterializedViewsCache.java]: done

File: ql/src/test/queries/clientpositive/materialized_view_rewrite_by_text_10.q
Comment (kasakrisz) [reply ql/src/test/queries/clientpositive/materialized_view_rewrite_by_text_10.q:7]: Yes. Creation should not fail but a message is logged and MV is disabled for rewrite if none of the algorithms are available.

In case of Alter the goal is to enable rewrite but if none of the algorithms are available I think it is ok to fail.

File: ql/src/test/queries/clientpositive/materialized_view_rewrite_by_text_11.q
Comment (kasakrisz) [reply ql/src/test/queries/clientpositive/materialized_view_rewrite_by_text_11.q]: I tried it. Let's see the result :)

File: ql/src/test/results/clientpositive/llap/materialized_view_no_cbo_rewrite.q.out
Comment (kasakrisz) [reply ql/src/test/results/clientpositive/llap/materialized_view_no_cbo_rewrite.q.out:39]: Actually these capabilities already exists if rewrite is not disabled at creation time. However Alter doesn't allow these. This patch tries to improve Alter too.

File: ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java
Comment (zabetak) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java:15942]: nit: Since we dropped "Automatic" from the result we can as well drop it from the methods since we are renaming them anyways.

Feel free to disregard or address directly during merge without waiting for a new CI run.

**Review by zabetak — APPROVED** [review]
LGTM!

**Comment by zabetak** [comment]
@kasakrisz By the way the test failure is relevant so make sure you address it before merging.

File: ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java
Comment (kasakrisz) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java]: Renamed. 
There were test failures in the last CI result due to a deprecated q out file. I have to rerun it anyway.

**Pull request merged**
